File: Parallel-Java/src/parallel/linq101/mylinq/MinMethods.java
package parallel.linq101.mylinq;

import org.javatuples.Pair;
import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Order;
import parallel.linq101.datasource.Product;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toMap;

public class MinMethods {
    //original:4 target:40
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }
    //min syntax : 5

    public int numbersMin() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int minNum = Arrays.stream(numbers).min().getAsInt();
        return minNum;
    }

    public int bigNumbersMin() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int minNum = Arrays.stream(numbers).filter(n -> n >= 5).min().getAsInt();
        return minNum;
    }

    public int oddNumbersMin() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int minNum = Arrays.stream(numbers).filter(n -> n % 2 == 0).min().getAsInt();
        return minNum;
    }

    public double doubleMin() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double minNum = Arrays.stream(numbers).min().getAsDouble();
        return minNum;
    }

    public double bigDoubleNumbersMin() {
        double[] numbers = {3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0};
        double minNum = Arrays.stream(numbers).filter(n -> n >= 5).min().getAsDouble();
        return minNum;
    }


    //min projection: 15

    public int numbersMinTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int minNum = Arrays.stream(numbers).map(n -> n * 2).min().getAsInt();
        return minNum;
    }

    public int bigNumbersMinTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int minNum = Arrays.stream(numbers).filter(n -> n >= 5).map(n -> n * 2).min().getAsInt();
        return minNum;
    }

    public int oddNumbersMinTwice() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        int minNum = Arrays.stream(numbers).filter(n -> n % 2 == 0).map(n -> n * 2).min().getAsInt();
        return minNum;
    }

    public int stringMinLength() {
        String[] words = {"cherry", "apple", "blueberry"};
        int minLength = Arrays.stream(words).mapToInt(w -> w.length()).min().getAsInt();
        return minLength;
    }

    public int shortStringMinLength() {
        String[] words = {"cherry", "apple", "blueberry"};
        int minLength = Arrays.stream(words).filter(w -> w.length() < 7).mapToInt(w -> w.length()).min().getAsInt();
        return minLength;
    }

    public double productUnitPriceMin() {
        List<Product> products = getProductList();
        double minUnitPrice = products.stream().mapToDouble(p -> p.getUnitPrice()).min().getAsDouble();
        return minUnitPrice;
    }

    public double expensiveProductUnitPriceMin() {
        List<Product> products = getProductList();
        double minUnitPrice = products.stream().filter(p -> p.getUnitPrice() > 100)
                .mapToDouble(p -> p.getUnitPrice()).min().getAsDouble();
        return minUnitPrice;
    }

    public int expensiveProductStockMin() {
        List<Product> products = getProductList();
        int minUnitPrice = products.stream().filter(p -> p.getUnitPrice() > 100)
                .mapToInt(p -> p.getUnitsInStock()).min().getAsInt();
        return minUnitPrice;
    }

    public int productLongNameLengthMin() {
        List<Product> products = getProductList();
        int minUnitPrice = products.stream().map(p -> p.getProductName())
                .filter(name -> name.length() > 7).mapToInt(name -> name.length()).min().getAsInt();
        return minUnitPrice;
    }

    public List<Pair<String, Double>> customerOrdersTotalMin() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalMin = customers.stream().map(c -> Pair.with(c.getCustomerId(),
                c.getOrders().stream().mapToDouble(o -> o.getTotal()).min().getAsDouble())).toList();
        return ordersTotalMin;
    }

    public Map<String, Double> customerOrdersTotalMinDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalMin = customers.stream().collect(toMap(c -> c.getCustomerId(),
                c -> c.getOrders().stream().mapToDouble(o -> o.getTotal()).min().getAsDouble()));
        return ordersTotalMin;
    }

    public List<Pair<String, Double>> customerInRegionOrdersTotalMin(String region) {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalMin = customers.stream().filter(c -> c.getRegion().equals(region))
                .map(c -> Pair.with(c.getCustomerId(),
                        c.getOrders().stream().mapToDouble(o -> o.getTotal()).min().getAsDouble())).toList();
        return ordersTotalMin;
    }

    public Map<String, Double> customerInRegionOrdersTotalMinDict(String region) {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalMin = customers.stream().filter(c -> c.getRegion().equals(region))
                .collect(toMap(c -> c.getCustomerId(),
                        c -> c.getOrders().stream().mapToDouble(o -> o.getTotal()).min().getAsDouble()));
        return ordersTotalMin;
    }

    public List<Pair<String, Double>> customerExpensiveOrdersTotalMin() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalMin = customers.stream().map(c -> Pair.with(c.getCustomerId(),
                c.getOrders().stream().filter(o -> o.getTotal() > 100)
                        .mapToDouble(o -> o.getTotal()).min().getAsDouble())).toList();
        return ordersTotalMin;
    }

    public Map<String, Double> customerExpensiveOrdersTotalMinDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalMin = customers.stream().collect(toMap(c -> c.getCustomerId(),
                c -> c.getOrders().stream().filter(o -> o.getTotal() > 100)
                        .mapToDouble(o -> o.getTotal()).min().getAsDouble()));
        return ordersTotalMin;
    }

    //min grouped: 10

    public List<Pair<String, Integer>> productEachCategoryStock() {
        List<Product> products = getProductList();
        List<Pair<String, Integer>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(p -> p.getUnitsInStock()).min().getAsInt())).toList();
        return categories;
    }

    public Map<String, Integer> productEachCategoryStockDict() {
        List<Product> products = getProductList();
        Map<String, Integer> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(p -> p.getUnitsInStock()).min().getAsInt()));
        return categories;
    }

    public List<Pair<String, Integer>> expensiveProductEachCategoryStock() {
        List<Product> products = getProductList();
        List<Pair<String, Integer>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().filter(
                        p -> p.getUnitPrice() > 100).mapToInt(p -> p.getUnitsInStock()).min().getAsInt())).toList();
        return categories;
    }

    public Map<String, Integer> expensiveProductEachCategoryStockDict() {
        List<Product> products = getProductList();
        Map<String, Integer> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().filter(
                        p -> p.getUnitPrice() > 100).mapToInt(p -> p.getUnitsInStock()).min().getAsInt()));
        return categories;
    }

    public List<Pair<String, Integer>> customerEachRegionOrdersMin() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Integer>> ordersMin = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(c -> c.getOrders().size()).min().getAsInt())).toList();
        return ordersMin;
    }

    public Map<String, Integer> customerEachRegionOrdersMinDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Integer> ordersMin = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(c -> c.getOrders().size()).min().getAsInt()));
        return ordersMin;
    }

    public List<Pair<String, Integer>> customerEachCountryOrdersMin() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Integer>> ordersMin = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().mapToInt(c -> c.getOrders().size()).min().getAsInt())).toList();
        return ordersMin;
    }

    public Map<String, Integer> customerEachCountryOrdersMinDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Integer> ordersMin = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().collect(toMap(g -> g.getKey(),
                        g -> g.getValue().stream().mapToInt(c -> c.getOrders().size()).min().getAsInt()));
        return ordersMin;
    }

    public List<Pair<String, Double>> customerEachRegionOrdersTotalMin() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, Double>> ordersTotalMin = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().flatMap(c ->
                        c.getOrders().stream()).mapToDouble(o -> o.getTotal()).min().getAsDouble())).toList();
        return ordersTotalMin;
    }

    public Map<String, Double> customerEachRegionOrdersTotalMinDict() {
        List<Customer> customers = getCustomerList();
        Map<String, Double> ordersTotalMin = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().flatMap(c ->
                        c.getOrders().stream()).mapToDouble(o -> o.getTotal()).min().getAsDouble()));
        return ordersTotalMin;
    }

    //min each group: 10
    public List<Pair<String, List<Product>>> productEachCategoryMinUnitPriceProducts() {
        List<Product> products = getProductList();
        List<Pair<String, List<Product>>> categories = products.stream().collect(groupingBy(p -> p.getCategory()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().collect(groupingBy(p -> p.getUnitPrice()))
                        .entrySet().stream().min(Map.Entry.comparingByKey()).get().getValue())).toList();
        return categories;
    }

    public Map<String, List<Product>> productEachCategoryMinUnitPriceProductsDict() {
        List<Product> products = getProductList();
        Map<String, List<Product>> categories = products.stream().collect(groupingBy(p -> p.getCategory())).entrySet()
                .stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().collect(groupingBy(
                        p -> p.getUnitPrice())).entrySet().stream().min(Map.Entry.comparingByKey()).get().getValue()));
        return categories;
    }

    public List<Pair<String, List<Product>>> expensiveProductEachCategoryMinUnitPriceProducts() {
        List<Product> products = getProductList();
        List<Pair<String, List<Product>>> categories = products.stream().filter(p -> p.getUnitPrice() > 100).
                collect(groupingBy(p -> p.getCategory())).entrySet().stream().map(g -> Pair.with(g.getKey(),
                        g.getValue().stream().collect(groupingBy(p -> p.getUnitPrice()))
                                .entrySet().stream().min(Map.Entry.comparingByKey()).get().getValue())).toList();
        return categories;
    }

    public Map<String, List<Product>> expensiveProductEachCategoryMinUnitPriceProductsDict() {
        List<Product> products = getProductList();
        Map<String, List<Product>> categories = products.stream().filter(p -> p.getUnitPrice() > 100).collect(
                groupingBy(p -> p.getCategory())).entrySet().stream().collect(toMap(g -> g.getKey(),
                g -> g.getValue().stream().collect(groupingBy(p -> p.getUnitPrice()))
                        .entrySet().stream().min(Map.Entry.comparingByKey()).get().getValue()));
        return categories;
    }

    public List<Pair<String, List<Customer>>> customerEachRegionOrdersMinCustomer() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, List<Customer>>> customerOrdersMin = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().
                        collect(groupingBy(c -> c.getOrders().size())).entrySet()
                        .stream().min(Map.Entry.comparingByKey()).get().getValue())).toList();
        return customerOrdersMin;
    }

    public Map<String, List<Customer>> customerEachRegionOrdersMinCustomerDict() {
        List<Customer> customers = getCustomerList();
        Map<String, List<Customer>> customerOrdersMin = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().
                        collect(groupingBy(c -> c.getOrders().size())).entrySet()
                        .stream().min(Map.Entry.comparingByKey()).get().getValue()));
        return customerOrdersMin;
    }

    public List<Pair<String, List<Customer>>> customerEachCountryOrdersMinCustomer() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, List<Customer>>> customerOrdersMin = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream().
                        collect(groupingBy(c -> c.getOrders().size())).entrySet()
                        .stream().min(Map.Entry.comparingByKey()).get().getValue())).toList();
        return customerOrdersMin;
    }

    public Map<String, List<Customer>> customerEachCountryOrdersMinCustomerDict() {
        List<Customer> customers = getCustomerList();
        Map<String, List<Customer>> customerOrdersMin = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream().
                        collect(groupingBy(c -> c.getOrders().size())).entrySet()
                        .stream().min(Map.Entry.comparingByKey()).get().getValue()));
        return customerOrdersMin;
    }

    public List<Pair<String, List<Order>>> customerEachRegionOrdersTotalMinCustomer() {
        List<Customer> customers = getCustomerList();
        List<Pair<String, List<Order>>> customerOrdersTotalMin = customers.stream().collect(groupingBy(c -> c.getRegion()))
                .entrySet().stream().map(g -> Pair.with(g.getKey(), g.getValue().stream()
                        .flatMap(c -> c.getOrders().stream()).collect(groupingBy(o -> o.getTotal()))
                        .entrySet().stream().min(Map.Entry.comparingByKey()).get().getValue())).toList();
        return customerOrdersTotalMin;
    }

    public Map<String, List<Order>> customerEachRegionOrdersTotalMinCustomerDict() {
        List<Customer> customers = getCustomerList();
        Map<String, List<Order>> customerOrdersTotalMin = customers.stream().collect(groupingBy(c -> c.getCountry()))
                .entrySet().stream().collect(toMap(g -> g.getKey(), g -> g.getValue().stream()
                        .flatMap(c -> c.getOrders().stream()).collect(groupingBy(o -> o.getTotal()))
                        .entrySet().stream().min(Map.Entry.comparingByKey()).get().getValue()));
        return customerOrdersTotalMin;
    }
}

File: Parallel-Csharp/src/Parallel/Linq101/MyLinq/MinMethods.cs
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101.MyLinq;

public class MinMethod {
    //original:4 target:40
    public List<Product> GetProductList() => Products.ProductList;
    public List<Customer> GetCustomerList() => Customers.CustomerList;

    //min syntax : 5

    public int NumbersMin() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int minNum = numbers.Min();
        return minNum;
    }

    public int BigNumbersMin() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int minNum = numbers.Where(n => n > 5).Min();
        return minNum;
    }

    public int OddNumbersMin() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int minNum = numbers.Where(n => n % 2 == 1).Min();
        return minNum;
    }

    public double DoubleMin() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double minNum = numbers.Min();
        return minNum;
    }

    public double BigDoubleNumbersMin() {
        double[] numbers = { 3.1, 4.2, 1.6, 0.7, 8.7, 6.9, 7.4, 2.5, 5.7, 0 };
        double minNum = numbers.Where(d => d > 5).Min();
        return minNum;
    }


    //min projection: 15

    public int NumbersMinTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int minNum = numbers.Min(n => n * 2);
        return minNum;
    }

    public int BigNumbersMinTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int minNum = numbers.Where(n => n > 5).Min(n => n * 2);
        return minNum;
    }

    public int OddNumbersMinTwice() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int minNum = numbers.Where(n => n % 2 == 1).Min(n => n * 2);
        return minNum;
    }

    public int StringMinLength() {
        string[] words = { "cherry", "apple", "blueberry" };
        int minLength = words.Min(w => w.Length);
        return minLength;
    }

    public int ShortStringMinLength() {
        string[] words = { "cherry", "apple", "blueberry" };
        int minLength = words.Where(w => w.Length < 7).Min(w => w.Length);
        return minLength;
    }

    public double ProductUnitPriceMin() {
        List<Product> products = GetProductList();
        double minUnitPrice = products.Min(p => p.UnitPrice);
        return minUnitPrice;
    }

    public double ExpensiveProductUnitPriceMin() {
        List<Product> products = GetProductList();
        double minUnitPrice = products.Where(p => p.UnitPrice > 100).Min(p => p.UnitPrice);
        return minUnitPrice;
    }

    public int ExpensiveProductStockMin() {
        List<Product> products = GetProductList();
        int minUnitPrice = products.Where(p => p.UnitPrice > 100).Min(p => p.UnitsInStock);
        return minUnitPrice;
    }

    public int ProductLongNameLengthMin() {
        List<Product> products = GetProductList();
        int minUnitPrice = products.Select(p => p.ProductName)
            .Where(name => name.Length > 7).Min(name => name.Length);
        return minUnitPrice;
    }

    public List<Tuple<string, double>> CustomerOrdersTotalMin() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalMin = customers.Select(c => Tuple.Create(c.CustomerId,
            c.Orders.Min(o => o.Total))).ToList();
        return ordersTotalMin;
    }

    public Dictionary<string, double> CustomerOrdersTotalMinDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalMin = customers.ToDictionary(c => c.CustomerId,
            c => c.Orders.Min(o => o.Total));
        return ordersTotalMin;
    }

    public List<Tuple<string, double>> CustomerInRegionOrdersTotalMin(string region) {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalMin = customers.Where(c => c.Region.Equals(region))
            .Select(c => Tuple.Create(c.CustomerId,
                c.Orders.Min(o => o.Total))).ToList();
        return ordersTotalMin;
    }

    public Dictionary<string, double> CustomerInRegionOrdersTotalMinDict(string region) {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalMin = customers.Where(c => c.Region.Equals(region))
            .ToDictionary(c => c.CustomerId,
                c => c.Orders.Min(o => o.Total));
        return ordersTotalMin;
    }

    public List<Tuple<string, double>> CustomerExpensiveOrdersTotalMin() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalMin = customers.Select(c => Tuple.Create(c.CustomerId,
            c.Orders.Where(o => o.Total > 100).Min(o => o.Total))).ToList();
        return ordersTotalMin;
    }

    public Dictionary<string, double> CustomerExpensiveOrdersTotalMinDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalMin = customers.ToDictionary(c => c.CustomerId,
            c => c.Orders.Where(o => o.Total > 100).Min(o => o.Total));
        return ordersTotalMin;
    }

    //min grouped: 10

    public List<Tuple<string, int>> ProductEachCategoryStock() {
        List<Product> products = GetProductList();
        List<Tuple<string, int>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, g.Min(p => p.UnitsInStock))).ToList();
        return categories;
    }

    public Dictionary<string, int> ProductEachCategoryStockDict() {
        List<Product> products = GetProductList();
        Dictionary<string, int> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g => g.Min(p => p.UnitsInStock));
        return categories;
    }

    public List<Tuple<string, int>> ExpensiveProductEachCategoryStock() {
        List<Product> products = GetProductList();
        List<Tuple<string, int>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key, g.Where(p => p.UnitPrice > 100)
                .Min(p => p.UnitsInStock))).ToList();
        return categories;
    }

    public Dictionary<string, int> ExpensiveProductEachCategoryStockDict() {
        List<Product> products = GetProductList();
        Dictionary<string, int> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key, g =>
                g.Where(p => p.UnitPrice > 100).Min(p => p.UnitsInStock));
        return categories;
    }

    public List<Tuple<string, int>> CustomerEachRegionOrdersMin() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, int>> ordersMin = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.Min(c => c.Orders.Length))).ToList();
        return ordersMin;
    }

    public Dictionary<string, int> CustomerEachRegionOrdersMinDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, int> ordersMin = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.Min(c => c.Orders.Length));
        return ordersMin;
    }

    public List<Tuple<string, int>> CustomerEachCountryOrdersMin() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, int>> ordersMin = customers.GroupBy(c => c.Country)
            .Select(g => Tuple.Create(g.Key,
                g.Min(c => c.Orders.Length))).ToList();
        return ordersMin;
    }

    public Dictionary<string, int> CustomerEachCountryOrdersMinDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, int> ordersMin = customers.GroupBy(c => c.Country)
            .ToDictionary(g => g.Key,
                g => g.Min(c => c.Orders.Length));
        return ordersMin;
    }

    public List<Tuple<string, double>> CustomerEachRegionOrdersTotalMin() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, double>> ordersTotalMin = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.SelectMany(c => c.Orders).Min(o => o.Total))).ToList();
        return ordersTotalMin;
    }

    public Dictionary<string, double> CustomerEachRegionOrdersTotalMinDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, double> ordersTotalMin = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.SelectMany(c => c.Orders).Min(o => o.Total));
        return ordersTotalMin;
    }

    //min each group: 10
    public List<Tuple<string, List<Product>>> ProductEachCategoryMinUnitPriceProducts() {
        List<Product> products = GetProductList();
        List<Tuple<string, List<Product>>> categories = products.GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key,
                g.Where(p => p.UnitPrice == g.Min(p => p.UnitPrice))
                    .ToList())).ToList();
        return categories;
    }
    
    public Dictionary<string, List<Product>> ProductEachCategoryMinUnitPriceProductsDict() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> categories = products.GroupBy(p => p.Category)
            .ToDictionary(g => g.Key,
                g=> g.Where(p => p.UnitPrice == g.Min(p => p.UnitPrice))
                    .ToList());
        return categories;
    }
    
    public List<Tuple<string, List<Product>>> ExpensiveProductEachCategoryMinUnitPriceProducts() {
        List<Product> products = GetProductList();
        List<Tuple<string, List<Product>>> categories = products.Where(p=>p.UnitPrice>100).GroupBy(p => p.Category)
            .Select(g => Tuple.Create(g.Key,
                g.Where(p => p.UnitPrice == g.Min(p => p.UnitPrice))
                    .ToList())).ToList();
        return categories;
    }
    
    public Dictionary<string, List<Product>> ExpensiveProductEachCategoryMinUnitPriceProductsDict() {
        List<Product> products = GetProductList();
        Dictionary<string, List<Product>> categories = products.Where(p=>p.UnitPrice>100).GroupBy(p => p.Category)
            .ToDictionary(g => g.Key,
                g=> g.Where(p => p.UnitPrice == g.Min(p => p.UnitPrice))
                    .ToList());
        return categories;
    }
    
    public List<Tuple<string, List<Customer>>> CustomerEachRegionOrdersMinCustomer() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, List<Customer>>> customerOrdersMin = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.Where(c => c.Orders.Length ==g.Min(c => c.Orders.Length)).ToList())).ToList();
        return customerOrdersMin;
    }

    public Dictionary<string, List<Customer>> CustomerEachRegionOrdersMinCustomerDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, List<Customer>> customerOrdersMin = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g => g.Where(c => c.Orders.Length ==g.Min(c => c.Orders.Length)).ToList());
        return customerOrdersMin;
    }

    public List<Tuple<string, List<Customer>>> CustomerEachCountryOrdersMinCustomer() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, List<Customer>>> customerOrdersMin = customers.GroupBy(c => c.Country)
            .Select(g => Tuple.Create(g.Key,
                g.Where(c => c.Orders.Length ==g.Min(c => c.Orders.Length)).ToList())).ToList();
        return customerOrdersMin;
    }

    public Dictionary<string, List<Customer>> CustomerEachCountryOrdersMinCustomerDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, List<Customer>> customerOrdersMin = customers.GroupBy(c => c.Country)
            .ToDictionary(g => g.Key,
                g => g.Where(c => c.Orders.Length ==g.Min(c => c.Orders.Length)).ToList());
        return customerOrdersMin;
    }

    public List<Tuple<string, List<Order>>> CustomerEachRegionOrdersTotalMinCustomer() {
        List<Customer> customers = GetCustomerList();
        List<Tuple<string, List<Order>>> customerOrdersTotalMin = customers.GroupBy(c => c.Region)
            .Select(g => Tuple.Create(g.Key,
                g.SelectMany(c=>c.Orders).Where(
                    o=>o.Total == g.SelectMany(c=>c.Orders).Min(o => o.Total)).ToList())).ToList();
        return customerOrdersTotalMin;
    }

    public Dictionary<string, List<Order>> CustomerEachRegionOrdersTotalMinCustomerDict() {
        List<Customer> customers = GetCustomerList();
        Dictionary<string, List<Order>> customerOrdersTotalMin = customers.GroupBy(c => c.Region)
            .ToDictionary(g => g.Key,
                g=>g.SelectMany(c => c.Orders).Where(
                    o => o.Total == g.SelectMany(c => c.Orders).Min(o => o.Total)).ToList());
        return customerOrdersTotalMin;
    }
}